src/core/server/ui_settings/ui_settings_service.ts

```diff
@@ -0,0 +1,140 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import { Observable } from 'rxjs';
+import { first } from 'rxjs/operators';
+import { CoreService } from '../../types';
+import { CoreContext } from '../core_context';
+import { Logger } from '../logging';
+
+import { SavedObjectsClientContract, SavedObjectAttribute } from '../saved_objects/types';
+import { HttpServiceSetup } from '../http';
+import { UiSettingsConfigType } from './ui_settings_config';
+import { IUiSettingsClient, UiSettingsClient } from './ui_settings_client';
+import { mapToObject } from '../../utils/';
+
+interface SetupDeps {
+  http: HttpServiceSetup;
+}
+
+/**
+ * UI element type to represent the settings.
+ * @public
+ * */
+export type UiSettingsType = 'json' | 'markdown' | 'number' | 'select' | 'boolean' | 'string';
+
+/**
+ * UiSettings parameters defined by the plugins.
+ * @public
+ * */
+export interface UiSettingsParams {
+  /** title in the UI */
+  name: string;
+  /** default value to fall back to if a user doesn't provide any */
+  value: SavedObjectAttribute;
+  /** description provided to a user in UI */
+  description: string;
+  /** used to group the configured setting in the UI */
+  category: string[];
+  /** a range of valid values */
+  options?: string[];
+  /** text labels for 'select' type UI element */
+  optionLabels?: Record<string, string>;
+  /** a flag indicating whether new value applying requires page reloading */
+  requiresPageReload?: boolean;
+  /** a flag indicating that value cannot be changed */
+  readonly?: boolean;
+  /** defines a type of UI element */
+  type?: UiSettingsType;
+}
+
+/** @internal */
+export interface UiSettingsServiceSetup {
+  /**
+   * Sets the parameters with default values for the uiSettings.
+   * @param values
+   */
+  setDefaults(values: Record<string, UiSettingsParams>): void;
```

Comment:
should we remove `defaults` term completely? plugins define not only values, but https://github.com/elastic/kibana/pull/47590/files#diff-39a763c69e856858c363b176d7490f48R46

Reply:
This makes sense to me :+1:

Reply:
I checked the code and realised that it's too much work right now to rename all the places using `defaults`. won't be done in this PR